src/stats/dendrogram.jl

```diff
@@ -0,0 +1,178 @@
+# Node struct
+struct DNode{N}
+    idx::Int
+    position::Point{N, Float32}
+    children::Union{Tuple{Int,Int}, Nothing}
+end
+
+function DNode(idx::Int, point::Point{N}, children::Union{Tuple{Int,Int}, Nothing}) where N
+    return DNode{N}(idx, point, children)
+end
+
+"""
+    dendrogram(x, y; kwargs...)
+
+Draw a [dendrogram](https://en.wikipedia.org/wiki/Dendrogram),
+with leaf nodes specified by `x` and `y` coordinates,
+and parent nodes identified by `merges`.
+# Arguments
+- `x`: x positions of leaf nodes
+- `y`: y positions of leaf nodes (default = 0)
+# Keywords
+- `merges`: specifies connections between nodes (see below)
+- `treestyle`: one of `:tree`, `:box`.  Overload `dendrogram_connectors(::Val{:mystyle}, parent, child1, child2)` to define a new style.
+"""
+@recipe(Dendrogram, nodes) do scene
+    Theme(
+        weights = Makie.automatic,
+        branch_shape = :box,
+        linewidth = Makie.inherit(scene, :linewidth, 1.0),
+        color = Makie.inherit(scene, :color, :black),
+        colormap = Makie.inherit(scene, :colormap, :tableau_10),
+        colorrange = Makie.automatic,
+        orientation = :vertical,
+        groups = nothing,
+        cycle = [:color => :patchcolor],
+        inspectable = Makie.inherit(scene, :inspectable, false),
+        xautolimits = Makie.inherit(scene, :xautolimits, true),
+        yautolimits = Makie.inherit(scene, :yautolimits, true),
+    )
+end
+
+function recursive_dendrogram_points(node, nodes, ret_points = Point2f[], ret_colors = []; color=:black, branch_shape=:tree, groups=nothing)
+    isnothing(node.children) && return nothing
+    child1 = nodes[node.children[1]]
+    child2 = nodes[node.children[2]]
+   
+    l = dendrogram_connectors(Val(branch_shape), node, child1, child2)
+    
+    # even if the inputs are 2d, the outputs should be 3d - this is what `to_ndim` does.
+    append!(ret_points, Makie.to_ndim.(Point3f, l, 0))
+    push!(ret_points, Point3f(NaN)) # separate segments
+
+    if isnothing(groups)
+        cgroup = 0
+    else
+        gs = recursive_leaf_groups(node, nodes, groups)
+        @debug gs
+        cgroup = length(unique(gs)) == 1 ? first(gs) : 0
+    end
+
+    @debug cgroup maxlog=2
+
+    append!(ret_colors, [cgroup for _ in 1:length(l)])
+    push!(ret_colors, NaN32) # separate segments
```

Comment:
Colors can be Float64 too

Reply:
grep'ing Makie/src i find an equal number of uses of Point3d and Point3f.  i had thought Float32 was preferred as that is what most consumer grade GPUs use.  is there a preference though for Float64 in the Makie code base?

Reply:
We recently switched to allow Float64 all the way through Makie, only converting when transferring data to the GPU in v0.21, just so we can keep the extra precision.  Especially if plotting over a small interval but with a high offset (e.g. `10^10:(10^10+10)` or something), you do end up needing Float64 precision.